File: LibTerra/Source/terradisc.cpp
#include "archestree.h"

/*!
\class TerraDisc archestree.h
\brief A disc primitive.
\ingroup ArchesTree
*/

/*!
\brief Create a disc primitive.
\param d Disc.
\param r Radius.
\param z Elevation.
\param a Alpha.
*/
TerraDisc::TerraDisc(const Disc2& d, const double& z, const double& r, const double& a) :TerraSkeleton(r, a), disc(d)
{
  TerraDisc::z = z;
  box666 = d.GetBox().Extended(r);
}

/*!
\brief Compute the fields.
\param p Point.
*/
ScalarAlpha TerraDisc::ElevationAlpha(const Vector2& p) const
{
  if (!box666.Inside(p))
    return ScalarAlpha();

  double d = disc.R(p);

  return ScalarAlpha(z, Alpha(d));
}

/*!
\brief Compute the size of a node.
*/
int TerraDisc::Memory() const
{
  return sizeof(TerraDisc);
}

/*!
\brief Renders the node into a QGraphicsScene.
\param scene The scene.
*/
void TerraDisc::Draw(QGraphicsScene& scene) const
{
  disc.Draw(scene);
  disc.Draw(scene);
  disc.Extended(r).Draw(scene);
}

File: Ptut_Translat/ptut2/TerraDisc.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Text;

namespace ptut2
{
	class TerraDisc : TerraSkeleton
	{
		Disc2 disc; //!< Disc.
		double z;     //!< Elevation.

		public TerraDisc(/*const*/ ref Disc2 d, /*const*/ ref double z, /*const*/ ref double r, /*const*/ ref double a):base(ref r,ref a)
		{
			this.disc = d;
			this.z = z;
			//box666 = d.GetBox().Extended(r);
		}

		public ScalarAlpha ElevationAlpha(/*const*/ ref Vector2 p) /*const*/
		{
			if (!box666.Inside(p))
				return new ScalarAlpha();

			double d = disc.R(p);

			return new ScalarAlpha(z, Alpha(d));
		}




	}
}
